### ISSUE TITLE
[ENH] Allow `ForecastingGridSearchCV` to return results with multiple score metrics

### ISSUE THREAD

**Issue by arnaujc91** [alan-turing-institute/sktime#6012]
**Is your feature request related to a problem? Please describe.**
I would like to have the same dataframe that currently `ForecastingGridSearchCV` returns as the attribute named `cv_results_ ` but on top of this i would like `ForecastingGridSearchCV` to evaluate the test set with multiple metrics as different metrics provide different information and currently `ForecastingGridSearchCV` just returns the metric used to rank the hyperparamters.

**Describe the solution you'd like**
Just return one column per metric in the dataframe. A similar approach is followed by [PyCaret](https://github.com/pycaret), e.g. 
https://github.com/pycaret/pycaret/blob/master/tutorials/time_series/forecasting/univariate_without_exogeneous_part2.ipynb

Look at the 10th output of the jupyter notebook.

![Screenshot 2024-02-26 at 13 56 21](https://github.com/sktime/sktime/assets/38285979/a5e14a54-63f1-4261-bd17-596daffb7548)

You should be able to pass a list of metrics and then also provide the information of which of the metrics are you going to use to rank the forecaster hyperparams.

**Comment by fkiraly** [comment]
This should not be too hard to add, given that internally, the `evaluate` benchmarking utility is used which already returns results of multiple metrics.

An open design question is by which metric to order the results and the tuning by - I think it should simply be the first metric in the list, if a list is passed.

**Comment by arnaujc91** [comment]
I would personally add a new argument in the API, sth like, `selected_metric` to make it really explicit. Picking automatically the first element of the list might lead to confusion I think. But this is just my taste, maybe someone else can also add their opinion here?

**Comment by fkiraly** [comment]
> Picking automatically the first element of the list might lead to confusion I think.

As long as it is well-described in the docstring?
Having two arguments seems like unnecessary multiplication of params to me - open to arguments though.

**Comment by sukjingitsit** [comment]
As a first-time contributor, I am interested in attempting to add this functionality.
Would it be useful if the user passes a list of params, of which by default it is the first element which gets chosen as the ranking metric, but the user has the ability to initialise a kwarg 'rank_from_last = False' which chooses the last element instead. Further, I believe this functionality will affect only the cv_results_ attribute which is derived from results (here the list of metrics would be used instead of scoring name.
![Screenshot 2024-02-27 at 13 09 06](https://github.com/sktime/sktime/assets/132339317/ea0517ba-ead4-4957-97b7-29e7ce0330a4)

**Comment by fkiraly** [comment]
> These are my arguments pro extra API argument

Yes, I think these are valid, assuming that if not a list but a single metric is passed, nothing changes.

Listing arguments against:

* having a conditional argument is less towards the "sensible defaults" design principle, and users would have to find and understand it, too, from the docstring.
* it is mathematically redundant, the same information can be conveyed with a single list.
* proliferation of arguments, the class already has about a hundred
* to avoid deprecation, the argument would have to be added at the end; moving it next to the metrics arg would require a full deprecation cycle

I'm not opposed to either solution though.

**Comment by arnaujc91** [comment]
Yeah for me the thing that annoys me the most is your last point actually... it is such an overkill. And yes i would keep the same functionality as now if a single metric is passed instead of a list. Would be great if someone else also shares their thoughts here :)

But I am also fine with your idea @fkiraly

**Comment by sukjingitsit** [comment]
As per the current state of the discussion section, I believe I can modify the ForecastingGridSearchCV and its base class to accommodate the following suggestions/guidelines,
1. evaluate the test set with multiple metrics as different metrics
2. passing the metrics as a list and default-choosing the first as the ranking metric
    1. allowing for an optional argument to specify whether ranking metric is first or last from the list (I feel any arbitrary choice would not be a good choice from a coding practice POV, and any user would choose either the first or last metric from their list for ranking.
3. maintaining the original non-list functionality using the original definitions

I would request someone go through my suggestion above and guide me further with respect to this issue. (I have also posted this on Discord, so I apologise if redundancy is an issue)

**Comment by fkiraly** [comment]
@sukjingitsit, yes, that would be nice - I agree with the suggestion, mostly.

point 2, sub-point i - I think having an arg to select only either first or last is confusing - I like @arnaujc91's suggestion on a selector argument better, and better than that I like the option of having no additional argument.

FYI @benHeid, @yarnabrina, any opinions from your side?

**Comment by sukjingitsit** [comment]
Hi, for the list, is this a good output style (user uses two metrics)
As for the ranks not being ascending, I noticed this is because ranks are ordered after cv_results_ is declared. I'd also like to change this, so that final cv_results_ is also rank ordered instead of it being according to the parameter grid as it currently is
![Screenshot 2024-02-28 at 10 00 39](https://github.com/sktime/sktime/assets/132339317/0f6ec473-2e9a-404a-90fb-983d342c94ec)

**Comment by fkiraly** [comment]
Hm, interesting.

* if it is not currently ordered by rank but by paramset, then let's keep that. We need to, to avoid breaking running code, and it makes sense in general.
* regarding table format, column order, column naming: optimally, the current order is a special case, and names are similar as the return of `evaluate`

**Comment by arnaujc91** [comment]
> Hi, for the list, is this a good output style (user uses two metrics) As for the ranks not being ascending, I noticed this is because ranks are ordered after cv_results_ is declared. I'd also like to change this, so that final cv_results_ is also rank ordered instead of it being according to the parameter grid as it currently is ![Screenshot 2024-02-28 at 10 00 39](https://private-user-images.githubusercontent.com/132339317/308421090-0f6ec473-2e9a-404a-90fb-983d342c94ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMTg2NzQsIm5iZiI6MTcwOTExODM3NCwicGF0aCI6Ii8xMzIzMzkzMTcvMzA4NDIxMDkwLTBmNmVjNDczLTJlOWEtNDA0YS05MGZiLTk4M2QzNDJjOTRlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyOFQxMTA2MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jOTY5YWZmZGY5YTBiMGI5N2M5MGUzZTE4NTc1MTUyY2I5NWE2NTRlZDcyMzAzZTcyMjk3NTRmZDAzNWVmNDFmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tluavCQEHEvzj_IQT2tuZTkbYCDdcPpT0ZSmoTUKoY4)

Just wondering but why do you have multiple forecasters in the same `cv_results_ `? Is that happening when you use the class `ForecastingGridSearchCV `? Or you just created it yourself as an example? 

Just confused because i thought `ForecastingGridSearchCV` just accepts one forecaster and not multiple.

**Comment by arnaujc91** [comment]
Btw the dataframe should be ordered by the metric ranking:

https://github.com/sktime/sktime/blob/02e3233f16f293e1cc476ece55200c52ff76bdd4/sktime/forecasting/model_selection/_tune.py#L262-L266

I understood is not the case @sukjingitsit

**Comment by fkiraly** [comment]
> that should not be changed

Yes, we need to be very careful that all code that currently runs does not change behaviour - or, we have to run through a deprecation cycle. The tuning classes are among the most widely used ones, so we ought to be extra careful.

**Comment by sukjingitsit** [comment]
To implement this, I was considering adding an optional parameter by the name of scoring_list in the base class (BaseGridSearch) which is otherwise set to None. This scoring_list will contain the name of the metrics the user wants to input (and may optionally contain the scoring metric, which is still separately declared under scoring). I am, however, facing one issue while using callables (as defined in sktime.performance_metrics.forecasting). They all get initialised as DynamicForecastingErrorMetric when .check is called, thus, when the .name parameter is called for naming the columns, they all return 'DynamicForecastingErrorMetric' which leads to duplicate values if more than one callable (causing Pandas errors). One simple resolution for this is to call .__name__ instead when .name is used and 'DynamicForecastingErrorMetric' is returned. However, I wanted to confirm if this approach is acceptable

**Comment by fkiraly** [comment]
Nice, thanks for the feedback and new ideas.

* Personally, I would go without an additional argument `scoring_list`, instead use `scoring`. I think it is worth hearing at least @yarnabrina, @benHeid, @arnaujc91 on this before making a definitive decision. My argument is, it would reduce the change footprint, and make it more easily generalizable
* I see the issue with callables. With the name, do you mena the name of the callable? That might work.

**Comment by arnaujc91** [comment]
I also think this is coming back to my first point, we seem to agree on adding no new arguments to the API. So I would stick with that or I would add a new argument to pick the forecaster for ranking but not an extra argument called `scoring_list `. The API should now accept either one metric or a list of metrics in the same argument as previously only accepting only a metric.

**Comment by sukjingitsit** [comment]
Please check the latest commit to the PR, there I have removed scoring_list, retaining scoring. If a list, scoring chooses the first metric specified as its ranking metric (can consider adding a separate param to choose which element of the list is to be used for ranking, but that has some issues : for instance naming and object issues may arise with callables, and an index based argument may lead to IndexError issues)

**Comment by arnaujc91** [comment]
Hi @sukjingitsit , did you open a PR? I did not see it

**Comment by sukjingitsit** [comment]
https://github.com/sktime/sktime/pull/6030, I have mentioned this issue for the PR

**Comment by arnaujc91** [comment]
> #6030, I have mentioned this issue for the PR

yeah sorry i missed it, the discussion is evolving quite fast hehe

**Comment by fkiraly** [comment]
a quick discussion on discord with @arnaujc91 has convinced me that his original suggestion is better than defaulting to first argument, i.e.:

* `scoring` allows a single metric or iterable of metric, or str keyed dict
* an additional argument `scoring_ranker: int` (or `str` if  `scoring` is `dict`) allows to select which metric to tune with, default = 0.

Why is this better? Because if you have multiple metrics and have prepared them as composites or with specific parameters etc, it is much easier to select `scoring_ranker`, for changing - or tuning it - than switcing the order of the list.

**Comment by yarnabrina** [comment]
> * Personally, I would go without an additional argument `scoring_list`, instead use `scoring`. I think it is worth hearing at least @yarnabrina, @benHeid, @arnaujc91 on this before making a definitive decision. My argument is, it would reduce the change footprint, and make it more easily generalizable

My personal preference is "explicit is better than implicit" and as a user I'll prefer a new argument. This can possibly be comparable to torch or tensorflow where you can track a set of metrics and for early stopping or other callbacks explicitly specify what to use. In fact in our office work based on top of sktime, I've already done this type of thing by two arguments: one mandatory to rank and one optional list to track others.

It seems I'm too late, but new argument has already been agreed upon, so that's great 👍

**Comment by sukjingitsit** [comment]
I will modify the method to add a specific param by the name of ranking_metric which will accept an integer for a list, or a string for a dictionary. For the dictionary part, are the dictionary keys only used for selecting the ranker? In this case, could there be an added restriction where the dictionary keys could be used as metric names if the user has specified a dictionary instead of a list

**Comment by arnaujc91** [comment]
> a quick discussion on discord with @arnaujc91 has convinced me that his original suggestion is better than defaulting to first argument, i.e.:
> 
> * `scoring` allows a single metric or iterable of metric, or str keyed dict
> * an additional argument `scoring_ranker: int` (or `str` if  `scoring` is `dict`) allows to select which metric to tune with, default = 0.
> 
> Why is this better? Because if you have multiple metrics and have prepared them as composites or with specific parameters etc, it is much easier to select `scoring_ranker`, for changing - or tuning it - than switcing the order of the list.

imo more robust than this is to pass `scoring` as type `List[BaseMetric]` and then `scoring_ranker` (type `string`) to be the name of the class of the metric u want to use to rank, which ofc has to be in the list. I think this is more robust because you do not need to make any assumptions about ordering of the list. Let's say you receive a list you just need to know that the metric you wanna use is there and that's all, no more details. Giving an integer i think is really prone to mistakes and you wont catch them, the code will run with the wrong metric.

**Comment by arnaujc91** [comment]
> > * Personally, I would go without an additional argument `scoring_list`, instead use `scoring`. I think it is worth hearing at least @yarnabrina, @benHeid, @arnaujc91 on this before making a definitive decision. My argument is, it would reduce the change footprint, and make it more easily generalizable
> 
> My personal preference is "explicit is better than implicit" and as a user I'll prefer a new argument. This can possibly be comparable to torch or tensorflow where you can track a set of metrics and for early stopping or other callbacks explicitly specify what to use. In fact in our office work based on top of sktime, I've already done this type of thing by two arguments: one mandatory to rank and one optional list to track others.
> 
> It seems I'm too late, but new argument has already been agreed upon, so that's great 👍

I also like your approach. But then the argument that we use right now `scoring` will have to be changed to `ranking_metric` or whatever new argument we agree on? Or you wanna keep using `scoring` for ranking and then pass an extra argument like `extra_metrics` (stupid name) where a user can pass a list such that we break the most minimum compatibility with the previous API?

**Comment by sukjingitsit** [comment]
So far, there are two possibilities,
1) We retain scoring as per its original functionality and also extend it to be the list/dict of metrics. Apart from this, the user can pass ranking_metric as a separate optional param to specify which element of the list/dict is to be the main metric, where default is first. (as per the current PR)
2) We retain scoring only as per its original functionality and make it the primary ranker. Apart from this, we initialise scoring_list for a list of secondary metrics only for ranking which may/may not contain the primary ranker (as per the first PR, which now needs some modifications)

Of these (1) makes more sense as it is not only more similar to the existing GridSearch but also much more convenient than passing the list/dict separately in my opinion

**Comment by arnaujc91** [comment]
I like both. Regarding the first approach, if a list of one element is passed or just one metric then the user doesn't have to pass any other argument to the API, as it is automatically selected. If the list contains more than one element, then we must force the user to pass an Argument. Do not select the first element by default please, thats what i argue is really prone to errors. The code might run with unwanted automatically selected first metric.

**Comment by sukjingitsit** [comment]
In that case, I'd suggest that instead of it raising an error if the list is more than one element and the user does not pass the ranking_metric argument (or an erroneous argument is passed instead), it autoselects the first value and outputs a warning along the lines of "GridSearchForecastCV has defaulted to the first argument, please specify 'ranking_metric' correctly to avoid this warning"

**Comment by fkiraly** [comment]
@sukjingitsit, I prefer option 1 and your suggestion with the warning (no exception) if `scoring` is a list and `ranking_metric` is not provided. Further, I prefer `ranking_metric` to allow int and str (str only if `scoring` is a `dict`), but not metrics.

**Comment by sukjingitsit** [comment]
I have updated the PR to reflect these changes, raising warnings and errors wherever needed

**Comment by sukjingitsit** [comment]
@fkiraly @yarnabrina @arnaujc91 please review the PR changes (I will add a code quality check PR after feedback is provided on the current commit, if needed)

**Comment by arnaujc91** [comment]
> @sukjingitsit, I prefer option 1 and your suggestion with the warning (no exception) if `scoring` is a list and `ranking_metric` is not provided. Further, I prefer `ranking_metric` to allow int and str (str only if `scoring` is a `dict`), but not metrics.

Don't you think that letting the user pass an integer, it being the position of the metric in the list, is really prone to errros? I mean you need to know the ordering of the list at that moment. Instead if you just pass the name of the metric (probably the class name), you do not have to really know anything about the ordering of the list. Also just from a code point of view:

```python
# scoring list is gotten through some magic above.
gscv = ForecastingGridSearchCV(scoring=scloring_list, ranking_metric=4)
# Ok i am debugging my code, what is the 4th element of the list scoring_list...?

# Versus
gscv = ForecastingGridSearchCV(scoring=scloring_list, ranking_metric="MeanSquaredError")
# I am debugging and here i do not care about ordering of the list, I know which metric I am using for ranking

```

As a developer in a project, i know very well what is happening in the second case. On the contrary if an integer is allowed and i am debugging i really need to figure out which is the 4th element in the passed list. Thats why i think is not a good idea.

**Comment by yarnabrina** [comment]
As I told above, I do use separate names only in my work. However I have a way to uniquely identify metrics because of higher level validation steps with pydantic, I'm not sure if sktime can do that easily or not.

For example mape and smape are derived from same class, how will you handle that with class names only?

**Comment by arnaujc91** [comment]
> As I told above, I do use separate names only in my work. However I have a way to uniquely identify metrics because of higher level validation steps with pydantic, I'm not sure if sktime can do that easily or not.
> 
> For example mape and smape are derived from same class, how will you handle that with class names only?

Good point, if they are the same class with different keyword arguments its challenging. Still i think is so much nicer to debug your code in the second case, but i have to think about how could that be done properly in the case you mentioned.

**Comment by fkiraly** [comment]
@arnaujc91, @sukjingitsit, @yarnabrina, I just checked how sklearn does it:

https://scikit-learn.org/stable/auto_examples/model_selection/plot_multi_metric_evaluation.html#sphx-glr-auto-examples-model-selection-plot-multi-metric-evaluation-py

In case of multi-metric evaluation, `refit` is used to select the metric - avoiding a new arg that way. I think that is better than all the options so far.

**Comment by arnaujc91** [comment]
Lol way better

**Comment by fkiraly** [comment]
Well, *sometimes* they do get designs right 😁

**Comment by sukjingitsit** [comment]
From what I understand, that is how we are using the ranking_metric parameter, at least when a dict is passed. The only differences arises when a list is passed. In sklearn, when a list is passed, it is restricted to string values and that would not lead to nomenclature issues as could be the case when using sktime (when passing callables or metrics of the same class). For the list case, we could make sure that we restrict scoring to a list of strings for an sklearn-like implementation, but that would limit the current applicability of sktime's gridsearch (in terms of how list arguments are passed to it)

**Comment by fkiraly** [comment]
> From what I understand, that is how we are using the ranking_metric parameter, at least when a dict is passed.

I see, @sukjingitsit, very astutely observed - we cannot do exactly the same as we have metrics that can be classes and have parameters.

I do not think we should restrict list to string values, hence.

How about allowing `refit` to be integer in addition?

Perhaps the `sklearn` design is not best, after all.

Should we keep the `ranking_metric` approach?

**Comment by yarnabrina** [comment]
Remembered an old discussion with @hazrulakmal where he pointed out a possible issue with using `refit`. Then users will not be able to specify whether to actually refit (train on full data ignoring cv splits) or not (only the model for latest split).

https://github.com/sktime/sktime/issues/3508#issuecomment-1589785727

**Comment by fkiraly** [comment]
Yes, indeed! That's a valid point.

**Comment by fkiraly** [comment]
FYI @hazrulakmal - are you still active on GitHub?

**Comment by sukjingitsit** [comment]
Hi, I went over the other thread on this issue and I'm not sure what the final conclusion is for the issue, as in how to deal with the list case (as I believe the dict case and base case are sufficiently addressed)?

**Comment by fkiraly** [comment]
I think the more recent discussion was inconclusive, which means the first consensus still holds:

* separate argument
* `scoring` can be str, metric, or list thereof, or str-keyed dict with values str or metric
* `ranking_metric` can be int, or str if `scoring` was dict

That is, I believe, what you implemented right now?
Apologies for the back/forth in specs.

**Comment by sukjingitsit** [comment]
Yes, that is what has been implemented so far. The current PR needs a code qual commit to deal with lines longer than 88 characters, else I believe it works fine. I will do so right now if that is the case (if code logic / implementation doesn't need to change)

**Comment by fkiraly** [comment]
👍 , I'd say let's wrap it up as-is in the interest of speed and not keeping the nice contribution hanging. The next release is in a few weeks anyway, and the logic can easily be changed between the variants discussed, from what you implemented - worst case (that I do not expect) reverted, or adapted by a core developer.

**Comment by hasan5352** [comment]
> **Is your feature request related to a problem? Please describe.** I would like to have the same dataframe that currently `ForecastingGridSearchCV` returns as the attribute named `cv_results_ ` but on top of this i would like `ForecastingGridSearchCV` to evaluate the test set with multiple metrics as different metrics provide different information and currently `ForecastingGridSearchCV` just returns the metric used to rank the hyperparamters.
> 
> **Describe the solution you'd like** Just return one column per metric in the dataframe. A similar approach is followed by [PyCaret](https://github.com/pycaret), e.g. https://github.com/pycaret/pycaret/blob/master/tutorials/time_series/forecasting/univariate_without_exogeneous_part2.ipynb
> 
> Look at the 10th output of the jupyter notebook.
> 
> ![Screenshot 2024-02-26 at 13 56 21](https://private-user-images.githubusercontent.com/38285979/307794149-a5e14a54-63f1-4261-bd17-596daffb7548.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwODMwNTYsIm5iZiI6MTcxMjA4Mjc1NiwicGF0aCI6Ii8zODI4NTk3OS8zMDc3OTQxNDktYTVlMTRhNTQtNjNmMS00MjYxLWJkMTctNTk2ZGFmZmI3NTQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE4MzIzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3ZDZlOWQ1MmE2YzEwYTY2NWMwMTE5NzRjN2M4YWUyZGFlY2EyYjY2YjU3YjBiMzIwZWIwNzdjNGYyNWVhMjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wvPJnQqmMjy20Z4JMNYwpWXh3tnH1n7byOH9zZemLOw)
> 
> You should be able to pass a list of metrics and then also provide the information of which of the metrics are you going to use to rank the forecaster hyperparams.

I have a high level plan to work on this:
1. Modify the ForecastingGridSearchCV class to calculate and store the evaluation metrics for each fold of cross-validation
2. Ensure that the dataframe returned by ForecastingGridSearchCV includes columns for each metric and each fold of cross-validation
3. Update the cv_results_ attribute of the ForecastingGridSearchCV class to include the newly calculated evaluation metrics.
4. Test the modified ForecastingGridSearchCV class to ensure that it returns the desired dataframe with multiple evaluation metrics for each fold of cross-validation

Is this plan correct? and can i work on this issue ?

**Comment by fkiraly** [comment]
@hasan5352, yes, that is a reasonable high-level plan.

There is a current PR here: https://github.com/sktime/sktime/pull/6030 - @sukjingitsit, are you still working on this?

If not, @hasan5352, you are welcome to start form scratch, or from the current state of #6030.
I would advise though, to avoid duplication, to wait a day or a few for @sukjingitsit to answer.

**Comment by sukjingitsit** [comment]
I am still working on the issue, I just need a few days due to some personal reasons. As mentioned earlier, the implementation is complete, I only have to ensure it passes all tests, and I am working on that part currently, since I do not fully understand which tests are failing and why

**Comment by fkiraly** [comment]
> since I do not fully understand which tests are failing and why

In such a case, you can of course simply ask for explanation on the PR.

**Comment by ksharma6** [comment]
Hi @sukjingitsit are you still working on this? I'd like to take over if you aren't available

**Comment by sukjingitsit** [comment]
Hey @ksharma6 you can take over this : ) The only main issue I am facing currently is the fact that to implement it, I am using lists, and that is causing an issue in the tests. fkiraly's comments on my commit seem to highlight the exact issue but I wasn't able to resolve it.

**Comment by fkiraly** [comment]
This is the PR: https://github.com/sktime/sktime/pull/6030

This should work already as implemented, except internally we should avoid indexing unhashable objects.

**Comment by ksharma6** [comment]
Hi [sukjingitsit](https://github.com/sukjingitsit) I hope you are doing well. I wanted to reach out and see if you would be able to give me a quick summary of where you left off on your PR (namely what method (s) were you working where you had to use lists). Any assistance would be greatly appreciated :)
